ls1intum/Artemis
# `Development`: Refactor repository access checks

### Checklist
#### General
#### Server
#### Changes affecting Programming Exercises

### Motivation and Context
The upcoming local version control system will require a way to check if a user is authorized to access a repository. The checks necessary for that are already implemented for the online editor.

### Description
This PR adds a `RepositoryAccessService` and refactors the access checks for the online editor to use that service. It can later also be used for authorizing local version control requests.
In a similar manner some code for notifying the user about a new result from the CI system is extracted from the `ResultResource` into a method in the `ProgrammingMessagingService`. It will later be used by the local CI system.

### Steps for Testing
Prerequisites:
- 1 Instructor
- 1 Student

1. Log in to Artemis as the instructor and create a new programming exercise.
2. Edit the source code for the different repositories (template, solution, tests) in the online editor and make sure all requests go through. Make sure the result is shown next to the submit button when building and testing is completed.
3. Log in as a student and participate in the programming exercise.
4. Edit your repository in the online editor and make sure you can only submit code before the due date.
5. Log in as an instructor again and create an exam with a programming exercise.
6. Log in as a student, participate in the exam and submit code using the online editor. You should not be able to submit code after the due date.

### Review Progress

#### Code Review
#### Manual Tests

### Test Coverage
| Class/File | Branch | Line |
|------------|-------:|-----:|
| RepositoryAccessService.java | 86% | 100% |
| ProgrammingSubmissionService.java | 76% | 94% |
| ProgrammingExerciseParticipationResource.java | 75% | 86% |
| ProgrammingSubmissionResource.java | 65% | 91% |
| ResultResource.java | 67% | 89% |
| RepositoryProgrammingExerciseParticipationResource.java | 70% | 79% |
| TestRepositoryResource.java | 66% | 81% |
| ProgrammingMessagingService.java | 75% | 100% |
